### DISCUSSION TITLE
Reduce Blocking time of Nuxt Applications

### DISCUSSION THREAD

**Discussion by asennoussi** [nuxt/nuxt#9061] [Questions (Nuxt 2)] [answered]
--Disclaimer--
Please excuse me if I make any conceptual mistakes. I'm a product guy but crazy about performance so I got myself into playing with code every now and then to test the performance of my app. 
In this ticket, I will use my product as a Benchmark and I'll be sharing essential information to see where the problem is.
I don't know if this problem is specific to my product or is a deficiency of how the framework works.

Ok, let's get started! 
My website is called [Mahkamati](https://mahkamaty.com/ar-ma) It's a Q2A for legal matters. 
On Web dev, it floats around [64 in performance](https://lighthouse-dot-webdotdevsite.appspot.com//lh/html?url=https%3A%2F%2Fmahkamaty.com%2Far-ma) which is not where I want it to be. 

On the Backend, we are using AdonisJS Proxied by Nginx and cached for 24 hours so the backend response is lightning fast ⚡️

But the problem we are facing right now, is the FE speed. 
On the FE, we use Nuxt JS+Vuetify + some plugins (Rudderstack, that injects FB pixel and others that are not that heavy)
As you can see from the Web dev results, our FID is high and the Total blocking time is high. 

The profiler looks like this x4 slowdown on a MAC (1.4 GHz Quad-Core Intel Core i5):
![image](https://user-images.githubusercontent.com/5939429/87161496-aa58f280-c2d5-11ea-9f01-971ed8ca1684.png)

You can download the profiler from here: 
[Profile-20200710T174938.txt](https://github.com/nuxt/nuxt.js/files/4903450/Profile-20200710T174938.txt)

As you can see, the main thread doesn't take a rest until later, so the user can not interact with the app. 

The direction where I wanted to go is explained by Google's [Web.dev](https://web.dev/lighthouse-total-blocking-time/)
Their idea is the following: 

If blocking time is anything above 50ms then instead of having a task that executes a JS for 2 secs, where TBT (Total blocking time) = 2000 - 50 = 1950 ms, you can split that file into 40 files where each executes at 50ms and your total blocking time will be from 1950 ms to 0ms! 
It's easier said than done, because lower size doesn't necessarily mean lower execution time. 
But, it's theoretically a good direction to take. 

So what causes this long JS task? 
If you look at the profiler, you can see 2 big JS executions, the second one is a big translation file (That's on me) but the first one is the app.js which is the entry point of the app, which includes all the modules and plugins that my app is using. 

Now my idea or question is, can we break down app.js into multiple tasks that execute around 50ms, how can we do it?
Is it Nuxt? Webpack? Both? 
I would love to hear your thoughts. 
Thank you.

**Comment by ricardobt** [comment]
Plus one here! We have exactly the same problem!
@sshuichi did you found any solution?

**Comment by iamsandeepkumarmandal** [comment]
Facing the same issue with my website.
https://www.rummytime.com/

**Comment by asennoussi** [comment]
@ricardobt, no. I'm hoping things will get better with Webpack5 and Vue3.

**Comment by voltane** [comment]
I'm guessing that this first blocking task is vue hydration process. With this plugin: https://github.com/maoberlehner/vue-lazy-hydration you can lower it down significantly (if properly used - read the specs).
Later on remember to render only required `app shell` and dynamically load other components (webpack will take care to split them into separate files).

**Comment by pi0** [comment, marked as answer]
I second @voltane idea to leverage lazy-hydration, reducing JS _execution time_ and Chunk splitting is mainly wepack's task so nothing much to do (hopefully would be improved by WP5)

One important tip worth mentioning is that nuxt plugins are usually main reason of blocking render/hydration since nuxt awaits on them before start rendering on client-side and also their dependencies will be added to main (or vendors) chunk which both are necessary to bootstrap too.

(NOTE: All examples below are for `.client` plugins)

**Example 1: Defer plugins with a background task**

Bad practice:

```js
export default async function(ctx, inject) {
   await ...
}
```

Doing task in parallel to render:

```js
async function task(ctx) {}

export default function(ctx, inject) {
   task(ctx).catch(console.error)
}
```

Using [`onNuxtReady`](https://nuxtjs.org/guides/concepts/context-helpers/#onnuxtready-helper) to defer task after app is mounted:

```js
async function task(ctx) {}

export default function(ctx, inject) {
   window.onNuxtReady(() => task(ctx).catch(console.error))
}
```

**Example 2: Lazy importing dependencies**

Bad practice:

```js
import bigDep from 'big-dep'

export default function (ctx, inject) {}
```

Create a chunk to reduce execution time (and also preload in parallel) but still blocks render:

```js
export default async function (ctx, inject) {
   const bigDep = await import('big-dep' /* webpackChunkName: 'big-dep' */) 
}
```

Do logic in background:

```js
async function task(ctx) {
   const bigDep = await import('big-dep' /* webpackChunkName: 'big-dep' */) 
}

export default function (ctx, inject) {
  task(ctx).catch(console.error)
}
```

**Example 3: Lazy import by usage:**

Bad practice:  (usage: `this.$util.foo()`)

```js
import getbigDep from 'big-dep'

export default function (ctx, inject) {
  inject('util', {
     async foo() {
          // some logic depending on bigDep
     }
  })
}
```

Lazy import by usage: (same usage)

```js
const getbigDep = () => import('big-dep' /* webpackChunkName: 'big-dep' */) 

const createUtils = ctx => ({
   async foo() {
      const bigDep = await getbigDep()
      // some logic depending on bigDep
  }
})

export default function (ctx, inject) {
  inject('util', createUtils(ctx))
}
```

Lazy import entire utils if utils themselfe are big: (Usage `this.$utils().then(utils => ...)`)

```js
// plugins/foo.utils.js
// we can either directly import or utilize lazy import too

export default (ctx) {
  return { ... }
}
```

```js
// plugins/foo.client.js
export default (ctx, inject) {
  inject(utils => import('./foo.utils').then(createUtils => createUtils(ctx))
}
```

**Comment by jd1378** [comment]
is this in the docs ? I haven't seen this info before, maybe add them to docs ? they were really helpful to me, I wasn't aware of it

**Comment by danielroe** [comment]
@jd1378 Docs PR covering this would definitely be welcome :heart:

**Comment by wusongliang** [comment]
Hope this article can help you
https://medium.com/@dario.ferderber/optimize-nuxt-js-website-for-google-pagespeed-insights-593c92be653d

**Comment by Envoy49** [comment]
@danielroe, @pi0  I've got TBT very high in my app using NUXT-BRIDGE, do you guys say that is not possible to reduce it due to how webpack works?
